src/exchange_price_source.ts

```diff
@@ -0,0 +1,158 @@
+import { ExchangeAdapter, Ticker } from './exchange_adapters/base'
+import BigNumber from 'bignumber.js'
+import { PriceSource, WeightedPrice } from './price_source'
+import { Exchange, OracleCurrencyPair } from './utils'
+import { MetricCollector } from './metric_collector'
+import { individualTickerChecks } from './aggregator_functions'
+
+export interface OrientedExchangePair {
+  exchange: Exchange
+  symbol: OracleCurrencyPair
+  toInvert: boolean
+}
+
+export interface ExchangePriceSourceConfig {
+  pairs: OrientedExchangePair[]
+}
+
+export type OrientedAdapter = {
+  adapter: ExchangeAdapter
+  toInvert: boolean
+}
+
+export type PairData = {
+  bid: BigNumber
+  ask: BigNumber
+  baseVolume: BigNumber
+  quoteVolume: BigNumber
+}
+
+function invertPair(pair: PairData): PairData {
+  return {
+    bid: pair.ask.exponentiatedBy(-1),
+    ask: pair.bid.exponentiatedBy(-1),
+    baseVolume: pair.quoteVolume,
+    quoteVolume: pair.baseVolume,
+  }
+}
+
+function tickerToPairData(ticker: Ticker): PairData {
+  return {
+    bid: ticker.bid,
+    ask: ticker.ask,
+    baseVolume: ticker.baseVolume,
+    // Using lastPrice to convert from baseVolume to quoteVolume, whereas
+    // ideally one would use the baseVolume execution's VWAP.
+    quoteVolume: ticker.baseVolume.multipliedBy(ticker.lastPrice),
+  }
+}
+
+async function fetchPairData(
+  adapter: OrientedAdapter,
+  maxPercentageBidAskSpread: BigNumber,
+  metricCollector?: MetricCollector
+): Promise<PairData> {
+  const ticker = await adapter.adapter.fetchTicker()
```

Comment:
`adapter.adapter` is a little confusing since the word means 2 slightly different things depending on its position in the chain:
- first `adapter` is an `OrientedAdapter` which is a container for an `ExchangeAdapter` and a `toInvert` boolean
- second `adapter` is an `ExchangeAdapter` capable of directly fetching data from an exchange

Perhaps the first `adapter` could be renamed to `orientedAdapter`?

Another, more complicated thought:
What about making `toInvert` a config value on the `ExchangeAdapter` class itself, and bring this inversion logic into `ExchangeAdapter`'s `fetchTicker()`?

Within an instance of a running oracle, would we ever need _both_ an inverted, and non-inverted version of the same adapter?

Reply:
Used object unpacking (and renamed to `orientedAdapter`). For now, the inversion logic remains on `PriceSource`.